### PR TITLE
Fix/opencl Better error Messages

### PR DESCRIPTION
**Pull request by GabrielZCode**
Better openCL error messages!! 

Using the same strategy as generate_nv() function in generate_stubs.sh , I've extracted the error messages from https://github.com/KhronosGroup/OpenCL-Headers/tree/main opencl Headers . --> generate_opencl_error_codes() at the same bash script with a new option   ./autogen_stubs.sh opencl-error-code   

Now by calling the check function on tinygrad/runtime/ops_gpu.py we get the original Error message from openCL , getting it by the statusCode .

The statusCode dict is located at tinygrad/runtime/autogen/opencl.py

### CODE REVIEW

File: tinygrad/runtime/ops_gpu.py [hunk 1]
```diff
@@ -5,12 +5,14 @@
...
 from tinygrad.device import BufferOptions, LRUAllocator, Compiled, Compiler, CompileError
+from tinygrad.runtime.autogen.opencl import opencl_status_codes
+
 
 # see test/external/external_osx_profiling.py to determine this ratio. it's in like GPU clocks or something
 OSX_TIMING_RATIO = (125/3) if OSX else 1.0
 
 def check(status):
-  if status != 0: raise RuntimeError(f"OpenCL Error {status}")
+    if status != 0: raise RuntimeError(f"OpenCL Error {opencl_status_codes.get(status, 'Unkown Error')}")
```
Comment (nimlgen) [inline tinygrad/runtime/ops_gpu.py:15]: spacing

File: .gitignore [hunk 2]
```diff
@@ -53,3 +53,9 @@ weights
 *.lprof
 *.pkl
 site/
+.clang-format
+.tool-versions
+setup_extra.py
```
Comment (nimlgen) [inline .gitignore:58]: unrelated change, why it's here?

File: autogen_stubs.sh [hunk 3]
```diff
@@ -22,12 +22,43 @@ fixup() {
   grep FIXME_STUB $1 || true
 }
 
+generate_opencl_error_codes(){
```
Comment (nimlgen) [inline autogen_stubs.sh:25]: no need to git clone, do the same way as it's done in `generate_opencl`. I think you can add just `cl_ext.h` there and parse errors. Also errors are negative values, can we just build a dict Dict[error int -> str] in ops_gpu with a one-liner?

File: .gitignore (hunk 2 above)
Comment (GabrielZCode) [reply .gitignore:58]: it's just some local files I don't want to push, if not in .gitignore , how should i make git ignore it

File: autogen_stubs.sh (hunk 3 above)
Comment (GabrielZCode) [reply autogen_stubs.sh:25]: are the other errors not important so i can just get what is present on cl_ext.h.
There's errors on cl.h, cl_egl.h, cl_dx9_media_sharing.h, cl_d3d11.h and cl_d3d10.h files.

About the one-liner, it's possible , it'll just not be genned by the bash script

**Comment by GabrielZCode** [comment]
I've tried other ways (looking for what symbols are exported from the openCL.so file , but the errors wasn't there), but using sed for extracting the errors was the simplest solution i've found, if anyone gets a better idea, i have no problems to implement it, but right now i don't have any.

File: autogen_stubs.sh (hunk 3 above)
Comment (GabrielZCode) [reply autogen_stubs.sh:25]: "do the same way as it's done in generate_opencl", you mean, just assume the header file it's already available on the local directory?, I don't get how that's better than git clone if the dir doesn't exist. OpenCL headers are not default to be present on all linux distros.

Comment (GabrielZCode) [reply autogen_stubs.sh:25]: cl_ext only have the -1141 error code, cl.h have 72 error codes

Comment (GabrielZCode) [reply autogen_stubs.sh:25]: just a note, there's two error codes that are not negative values, they are hexadecimal [0x40D4] [0x40D5]

File: tinygrad/runtime/ops_gpu.py [hunk 4]
```diff
@@ -5,12 +5,14 @@
...
 from tinygrad.device import BufferOptions, LRUAllocator, Compiled, Compiler, CompileError
+from tinygrad.runtime.autogen.opencl import opencl_status_codes
+
 
 # see test/external/external_osx_profiling.py to determine this ratio. it's in like GPU clocks or something
 OSX_TIMING_RATIO = (125/3) if OSX else 1.0
 
 def check(status):
-  if status != 0: raise RuntimeError(f"OpenCL Error {status}")
+  if status != 0: raise RuntimeError(f"OpenCL Error {opencl_status_codes.get(status, 'Unkown Error')}")
```
Comment (nimlgen) [inline tinygrad/runtime/ops_gpu.py:15]: typo in `Unknown`

Comment (GabrielZCode) [reply tinygrad/runtime/ops_gpu.py]: fixed

File: autogen_stubs.sh (hunk 3 above)
Comment (GabrielZCode) [reply autogen_stubs.sh:25]: @nimlgen can you help me on the questions above? Any other thing missing for the PR merge?

File: .gitignore (hunk 2 above)
Comment (geohot) [reply .gitignore:58]: This is not resolved

Comment (GabrielZCode) [reply .gitignore]: ok. fixed

File: tinygrad/runtime/ops_gpu.py [hunk 5]
```diff
@@ -5,12 +5,14 @@
...
 from tinygrad.device import BufferOptions, LRUAllocator, Compiled, Compiler, CompileError
+from tinygrad.runtime.autogen.opencl import opencl_status_codes
+
 
 # see test/external/external_osx_profiling.py to determine this ratio. it's in like GPU clocks or something
 OSX_TIMING_RATIO = (125/3) if OSX else 1.0
 
 def check(status):
-  if status != 0: raise RuntimeError(f"OpenCL Error {status}")
+  if status != 0: raise RuntimeError(f"OpenCL Error {opencl_status_codes.get(status, 'Unknown Error')}")
```
Comment (chenyuxyz) [inline tinygrad/runtime/ops_gpu.py:15]: i think it at least should print the status code in the unknown case.

Comment (GabrielZCode) [reply tinygrad/runtime/ops_gpu.py]: ok. done

Comment (GabrielZCode) [reply tinygrad/runtime/ops_gpu.py]: it'll print in all cases, it's a good information to have

**Comment by GabrielZCode** [comment]
Simplified a lot the gen_stubs function, now it's just two seds and two cats commands, the rest is the git clone check (only run if the directory does not exist), the 100 lines is the error dict in opencl.py

File: autogen_stubs.sh (hunk 3 above)
Comment (GabrielZCode) [reply autogen_stubs.sh:25]: @nimlgen i think i got your idea, did a simplification of the seds

File: tinygrad/runtime/autogen/opencl.py [hunk 6]
```diff
@@ -1793,3 +1793,107 @@ class struct__cl_name_version(Structure):
     'struct__cl_kernel', 'struct__cl_mem', 'struct__cl_name_version',
     'struct__cl_platform_id', 'struct__cl_program',
     'struct__cl_sampler', 'union__cl_image_desc_0']
+opencl_status_codes = {}
+opencl_status_codes[-1002] = "CL_INVALID_D3D10_DEVICE_KHR                         "
```
Comment (nimlgen) [inline tinygrad/runtime/autogen/opencl.py:1797]: all strings have a lot of spaces at the end

File: autogen_stubs.sh [hunk 7]
```diff
@@ -168,6 +195,7 @@ generate_hsa() {
 }
 
 if [ "$1" == "opencl" ]; then generate_opencl
+elif [ "$1" == "opencl-error-code" ]; then generate_opencl_error_codes
```
Comment (nimlgen) [inline autogen_stubs.sh:198]: move into opencl

File: autogen_stubs.sh [hunk 8]
```diff
@@ -22,12 +22,38 @@ fixup() {
...
+  if [ ! -d "$OPENCL_HEADERS_SRC" ]; then
+    git clone https://github.com/KhronosGroup/OpenCL-Headers.git $OPENCL_HEADERS_SRC
+    pushd .
+    cd $OPENCL_HEADERS_SRC
+    git reset --hard $OPENCL_COMMIT_HASH
+    popd
+  fi
+  
+  file=$OPENCL_HEADERS_SRC/opencl_headers_coalesce.h
+  cat $OPENCL_HEADERS_SRC/CL/*.h > $file # Form a single file from all headers, to extract all errors present
```
Comment (nimlgen) [inline autogen_stubs.sh:38]: in `opencl` we take headers from `/usr/include/CL`, so let's not change this *partially* in this pr, can you take all needed headers from `/usr/include/CL/`?

Comment (GabrielZCode) [reply autogen_stubs.sh:38]: So, I think that's a distro compatibility problem, for example, I'm running openSuse and /usr/include doesn't even have a CL folder, of course I can fix it and put it myself, but other developers might run on the same issue.

If that's acceptable , i can comment it (for reference) and set the OPENCL_HEADERS_SRC=/usr/include/CL .

Should i keep it as is or force the /usr/include/CL usage?

File: autogen_stubs.sh (hunk 7 above)
Comment (GabrielZCode) [reply autogen_stubs.sh]: done

File: tinygrad/runtime/autogen/opencl.py (hunk 6 above)
Comment (GabrielZCode) [reply tinygrad/runtime/autogen/opencl.py:1797]: fixed, but it's in the fork now that the PR is closed
